app/src/main/java/com/x8bit/bitwarden/data/platform/datasource/disk/KeyDiskSourceImpl.kt

```
@@ -0,0 +1,190 @@
+                }
+            }
+
+        // Step 2: Get a list of aliases and choose the first one.
+        val internalAlias = pkcs12KeyStore.aliases()
+            ?.takeIf { it.hasMoreElements() }
+            ?.nextElement()
+            ?: return ImportPrivateKeyResult.Error.UnsupportedKey
+
+        // Step 3: Extract PrivateKey and X.509 certificate from the KeyStore and verify
+        // certificate alias.
+        val privateKey = try {
+            pkcs12KeyStore.getKey(internalAlias, password.toCharArray())
+                ?: return ImportPrivateKeyResult.Error.UnrecoverableKey
+        } catch (e: UnrecoverableKeyException) {
+            Timber.e(e, "Failed to get private key")
+            return ImportPrivateKeyResult.Error.UnrecoverableKey
+        }
+
+        val certChain: Array<Certificate> = pkcs12KeyStore
+            .getCertificateChain(internalAlias)
+            ?.takeUnless { it.isEmpty() }
+            ?: return ImportPrivateKeyResult.Error.InvalidCertificateChain
+
+        // Step 4: Store the private key and X.509 certificate in the AndroidKeyStore if the alias
+        // does not exists.
+        with(androidKeyStore) {
+            if (containsAlias(alias)) {
+                return ImportPrivateKeyResult.Error.DuplicateAlias
+            }
+
+            try {
+                setKeyEntry(alias, privateKey, null, certChain)
+            } catch (e: KeyStoreException) {
+                Timber.e(e, "Failed to import key into Android KeyStore")
+                return ImportPrivateKeyResult.Error.KeyStoreOperationFailed
+            }
+        }
+        return ImportPrivateKeyResult.Success(alias)
+    }
+
+    override fun removeMutualTlsKey(
+        alias: String,
+        host: MutualTlsKeyHost,
+    ): RemoveMutualTlsKeyResult =
+        when (host) {
+            MutualTlsKeyHost.ANDROID_KEY_STORE -> removeKeyFromAndroidKeyStore(alias)
+            MutualTlsKeyHost.KEY_CHAIN -> RemoveMutualTlsKeyResult.Success
```

❓ I am assuming that this would be for a cert manually added to the devices through the security settings, which we would not be able to remove. If that is correct I am guessing there is some handling up stream like maybe redirecting them to the settings or something? I think that this makes sense to have `MutualTlsKeyHost` if there are new ones we need to support in the future. I think my question is what are we using this result for and would it make sense to have a like a `NotMyProblem` result type?

---

Correct. `KEY_CHAIN` represents keys installed via system settings and we cannot remove them from within the app.

Now that you ask, I don't think we need this result. At least not right now. At the moment, this will only ever be called when the user logs out, so we wouldn't want to stop that process if removal of the key fails. It may be better to ignore any errors that occur. 🤔 

If/when we add proper key management UI within the app, linking to the system settings for system certificates may be something we do, and it would also create a need for the result.

I'll remove the result for now. We can add it back later if it's needed.